src/Server/HTTPHandler.cpp

```diff
@@ -364,16 +364,26 @@ bool HTTPHandler::authenticateUser(
     /// The header 'X-ClickHouse-SSL-Certificate-Auth: on' enables checking the common name
     /// extracted from the SSL certificate used for this connection instead of checking password.
     bool has_ssl_certificate_auth = (request.get("X-ClickHouse-SSL-Certificate-Auth", "") == "on");
-    bool has_auth_headers = !user.empty() || !password.empty() || !quota_key.empty() || has_ssl_certificate_auth;
+    bool has_auth_headers = !user.empty() || !password.empty() || has_ssl_certificate_auth;
 
     /// User name and password can be passed using HTTP Basic auth or query parameters
     /// (both methods are insecure).
     bool has_http_credentials = request.hasCredentials();
-    bool has_credentials_in_query_params = params.has("user") || params.has("password") || params.has("quota_key");
+    bool has_credentials_in_query_params = params.has("user") || params.has("password");
 
     std::string spnego_challenge;
     std::string certificate_common_name;
 
+    if (params.has("quota_key"))
+    {
+        if (!quota_key.empty())
+            throw Exception(ErrorCodes::AUTHENTICATION_FAILED,
+                            "Invalid authentication: it is not allowed "
+                            "to use quota key as HTTP header and quota key as parameter simultaneously");
+
+        quota_key = params.get("quota_key");
+    }
```

Comment:
I think it's better to move this quota-specific code later near the place `quota_key` is used.
Also it seems the error code here should be rather something like `BAD_ARGUMENTS`.